### ISSUE TITLE
Add a stroke-dahcap option to LineSymbolizer.

### ISSUE THREAD

**Issue by StyXman** [mapnik/mapnik#3982]
Mapnik has the `stroke-linecap` option that handles how ends of lines are rendered. One of the consequences is that when you use `stroke-dasharray`, the same cap is used for the dashes. I would like to define the dash caps independently.

I also plan to try to implement this myself, so if you could just give me some pointer where to start, I would really appreciate it.

**Comment by talaj** [comment]
That would be great, @StyXman . I see that you already opened this issue in https://github.com/mapnik/mapnik/issues/3849. Can you please close it?

Concerning implementation, I can offer some basic help.

There is a pipeline of operations which geometry goes through. These operations convert a stream of vertices of the geometry to another stream of vertices. Here is a list of operations for LineSymbolizer:

https://github.com/mapnik/mapnik/blob/2929c4ae6db6d6c47eff9789bc5043d57ff3dc70/src/agg/process_line_symbolizer.cpp#L187-L191

[`vertex_converter`](https://github.com/mapnik/mapnik/blob/master/include/mapnik/vertex_converters.hpp) is a class which chains those operations. 

The last two from the list are [dash](https://github.com/mapnik/mapnik/blob/2929c4ae6db6d6c47eff9789bc5043d57ff3dc70/include/mapnik/vertex_converters.hpp#L117-L140) and [stroke](https://github.com/mapnik/mapnik/blob/2929c4ae6db6d6c47eff9789bc5043d57ff3dc70/include/mapnik/vertex_converters.hpp#L176-L195) converters. Dash converter splits line to dashes - many short lines. Stroke converter converts lines to polygons with given parameters (such as `stroke-width` and `stroke-linecap`). Resulting polygons are rasterized.

Your target is therefore the stroke converter. I would try to extend it or rather introduce specialized stroke converter for dashed lines to keep original code efficient.

The dash converter and the stroke converter are part of AGG library as templates [`agg::conv_stroke`](https://github.com/mapnik/mapnik/blob/master/deps/agg/include/agg_conv_stroke.h) and [`agg::conv_dash`](https://github.com/mapnik/mapnik/blob/master/deps/agg/include/agg_conv_dash.h). It may be useful to read [Introduction to AGG library](http://www.antigrain.com/doc/introduction/introduction.agdoc.html). There are some [Demo Examples](http://www.antigrain.com/demo/index.html), especially `conv_stroke.cpp` is useful for our case.

**Comment by StyXman** [comment]
Yeah, when I was filling the info for the issue I had a déjà-vu sensation :)

Thanks for the pointers. I'll try to do it, but I really don't have much time.

**Comment by StyXman** [comment]
It took me a long time to have the time to look at this with time.

Could you explain how does the rendering machinery work? From what I read, the converters are instantiated, then vertexes are added with `add_vertex()` and then `vertex()` is called, which runs a quasi-state-machine (they keep track of past and future states) to do... what? 

What I'm thinking of is to extend the dasher to mark the first and last segments of the original line, and the stroker to use that to use one cap or another, but I can't see how the data is fed to the converters, and how the converters modify the data so it's fed to the following one. I think I see the hooks, but I need to understand the machinery that makes them tick, and AGG's site is not really explicit about it. i didn't manage to extract that info form the example you gave me.
